File: assets/map/elements/item/sword/sword.ts
const scriptPath = Script.getInfo().path;

type Idle = { status: "idle" };
type Swinging = { status: "swinging"; frame: number };
type Cooldown = { status: "cooldown"; frame: number };
type ItemState = Idle | Swinging | Cooldown;
const itemStateInit: ItemState = { status: "idle" };

const COOLDOWN_FRAMES = 15;
const ATTACK_FPS = 10;

export default {
  preUpdateInGame() {
    // Hydrate newly spawned sword items
    const names = world.query(EntityName);
    const items = world.query(Item);
    for (const { entity, components } of items) {
      const [item] = components;

      // If this is one of our items without a name
      if (item.script == scriptPath && !names.get(entity)) {
        // Hydrate the entity
        world.insert(entity, Value.create(EntityName, ["Item: Sword"]));

        // Add the animated sprite
        world.insert(
          entity,
          Value.create(AnimatedSprite, {
            start: 0,
            end: 0,
            repeat: false,
            fps: ATTACK_FPS,
            atlas: {
              id: Assets.getHandleId("sword.atlas.yaml"),
            },
          })
        );
        // And the kinematic body
        world.insert(
          entity,
          Value.create(KinematicBody, {
            size: {
              x: 64,
              y: 16,
            },
            offset: {
              y: 38,
            },
            gravity: 1,
            has_friction: true,
            has_mass: true,
          })
        );
      }
    }
  },

  updateInGame() {
    const players = world.query(AnimatedSprite, Transform, PlayerIdx);
    const parents = world.query(Parent);
    const items = world.query(
      Transform,
      KinematicBody,
      AnimatedSprite,
      GlobalTransform,
      Item
    );

    // Helper to spawn a damage region
    const spawnDamageRegion = (
      owner: Entity,
      x: number,
      y: number,
      width: number,
      height: number
    ) => {
      /// This is a hack to get a global transform because scripts can't construct it with
      /// Value.create(). ( Fixed in Bevy 0.9 )
      const globalTransform = items[0].components[3];

      // Spawn damage region entity
      let entity = world.spawn();
      world.insert(
        entity,
        Value.create(Transform, {
          translation: {
            x,
            y,
          },
        })
      );
      world.insert(entity, globalTransform);
      world.insert(
        entity,
        Value.create(DamageRegion, {
          size: {
            x: width,
            y: height,
          },
        })
      );
      world.insert(entity, Value.create(DamageRegionOwner, [owner]));
      world.insert(
        entity,
        Value.create(Lifetime, {
          lifetime: 1 / 60,
        })
      );
    };

    // Update items that are being held
    //
    // This section will make the item follow the player around and match the player's facing
    // direction.
    for (const { entity: itemEnt, components } of items) {
      const [itemTransform, body, sprite] = components;
      const state = Script.getEntityState<ItemState>(itemEnt, itemStateInit);

      let parentComponents = parents.get(itemEnt);
      // If this item isn't being held, skip the item
      if (!parentComponents) continue;

      const [parent] = parentComponents;
      const [playerSprite, playerTransform] = players.get(parent[0]);

      // Deactivate item collision
      body.is_deactivated = true;

      // Set animation to default position if we aren't swinging
      if (state.status != "swinging") {
        sprite.start = 4;
        sprite.end = 4;
        sprite.index = 0;
        sprite.repeat = false;
      }

      // Flip the sprite to match our player orientation
      const flip = playerSprite.flip_x;
      sprite.flip_x = flip;
      const flipFactor = flip ? -1 : 1;
      // Align the sprite with the player's position
      itemTransform.translation = Value.create(Vec3, {
        x: 13 * (flip ? -1 : 1),
        y: 21,
      });

      // If we're swinging the weapon
      if (state.status == "swinging") {
        // If we are at the end of the swing animation
        if (sprite.index >= sprite.end - sprite.start - 1) {
          // Go to cooldown frames
          Script.setEntityState<ItemState>(itemEnt, {
            status: "cooldown",
            frame: 0,
          });

          // Set the current attack frame to the animation frame index
        } else {
          state.frame = sprite.index;
        }

        // Trigger frame collisions for each sword animation position
        switch (state.frame) {
          case 0:
            spawnDamageRegion(
              parent[0],
              playerTransform.translation.x + 20 * flipFactor,
              playerTransform.translation.y + 20,
              30,
              70
            );
          case 1:
            spawnDamageRegion(
              parent[0],
              playerTransform.translation.x + 25 * flipFactor,
              playerTransform.translation.y + 20,
              40,
              50
            );
          case 2:
            spawnDamageRegion(
              parent[0],
              playerTransform.translation.x + 20 * flipFactor,
              playerTransform.translation.y,
              40,
              50
            );
        }

        state.frame += 1;

        // If we are in cooldown frames
      } else if (state.status == "cooldown") {
        // If cooldown frames have finished
        if (state.frame >= COOLDOWN_FRAMES) {
          // Go back to idle state
          Script.setEntityState<ItemState>(itemEnt, { status: "idle" });
        } else {
          state.frame += 1;
        }
      }
    }

    // For every item that is being used
    for (const event of Items.useEvents()) {
      // Get the current item state
      const state = Script.getEntityState<ItemState>(event.item, itemStateInit);

      if (state.status == "idle") {
        const [_itemTransform, _body, sprite] = items.get(event.item);

        // Start attacking animation
        sprite.index = 0;
        sprite.start = 8;
        sprite.end = 12;

        // And move to an attacking state
        Script.setEntityState<ItemState>(event.item, {
          status: "swinging",
          frame: 0,
        });
      }
    }

    // Update dropped items
    for (const event of Items.dropEvents()) {
      const [item_transform, body, sprite] = items.get(event.item);

      // Re-activate physics body on the item
      body.is_deactivated = false;
      // Put sword in rest position
      sprite.start = 0;
      sprite.end = 0;
      // Make sure item maintains player velocity
      body.velocity = event.velocity;
      body.is_spawning = true;

      // Drop item at the middle of the player
      item_transform.translation.y = event.position.y - 30;
      item_transform.translation.x = event.position.x;
      item_transform.translation.z = event.position.z;
    }
  },
};

File: src/map/elements/sword.rs
use crate::player::PlayerKillCommand;

use super::*;

pub struct SwordPlugin;
impl Plugin for SwordPlugin {
    fn build(&self, app: &mut App) {
        app.add_rollback_system(RollbackStage::PreUpdate, pre_update_in_game)
            .add_rollback_system(RollbackStage::Update, update_in_game)
            .extend_rollback_plugin(|plugin| plugin.register_rollback_type::<Sword>());
    }
}

#[derive(Component, Reflect, Default, Clone)]
#[reflect(Component, Default)]
pub struct Sword {
    pub state: SwordState,
    dropped_time: f32,
}

#[derive(Component, Reflect, Default, Clone)]
#[reflect(Component, Default)]
pub enum SwordState {
    #[default]
    Idle,
    Swinging {
        frame: usize,
    },
    Cooldown {
        frame: usize,
    },
}

const COOLDOWN_FRAMES: usize = 13;
const ATTACK_FPS: f32 = 10.0;

fn pre_update_in_game(
    mut commands: Commands,
    non_hydrated_map_elements: Query<
        (Entity, &Sort, &Handle<MapElementMeta>, &Transform),
        Without<MapElementHydrated>,
    >,
    mut ridp: ResMut<RollbackIdProvider>,
    element_assets: Res<Assets<MapElementMeta>>,
) {
    // Hydrate any newly-spawned swords
    let mut elements = non_hydrated_map_elements.iter().collect::<Vec<_>>();
    elements.sort_by_key(|x| x.1);
    for (entity, _sort, map_element_handle, transform) in elements {
        let map_element = element_assets.get(map_element_handle).unwrap();

        if let BuiltinElementKind::Sword {
            atlas_handle,
            can_rotate,
            ..
        } = &map_element.builtin
        {
            commands.entity(entity).insert(MapElementHydrated);

            commands
                .spawn()
                .insert(Rollback::new(ridp.next_id()))
                .insert(Item {
                    script: "core:sword".into(),
                })
                .insert(Name::new("Item: Sword"))
                .insert(Sword::default())
                .insert(AnimatedSprite {
                    start: 0,
                    end: 0,
                    atlas: atlas_handle.inner.clone(),
                    repeat: false,
                    fps: ATTACK_FPS,
                    ..default()
                })
                .insert_bundle(VisibilityBundle::default())
                .insert(MapRespawnPoint(transform.translation))
                .insert_bundle(TransformBundle {
                    local: *transform,
                    ..default()
                })
                .insert(map_element_handle.clone())
                .insert(KinematicBody {
                    size: Vec2::new(64.0, 16.0),
                    gravity: 1.0,
                    has_mass: true,
                    has_friction: true,
                    can_rotate: *can_rotate,
                    ..default()
                });
        }
    }
}

fn update_in_game(
    mut commands: Commands,
    players: Query<(&AnimatedSprite, &Transform, &KinematicBody), With<PlayerIdx>>,
    mut swords: Query<
        (
            Entity,
            &mut Transform,
            &mut Sword,
            &mut AnimatedSprite,
            &mut KinematicBody,
            &Handle<MapElementMeta>,
            Option<&Parent>,
            Option<&ItemUsed>,
            Option<&ItemDropped>,
        ),
        Without<PlayerIdx>,
    >,
    mut ridp: ResMut<RollbackIdProvider>,
    player_inputs: Res<PlayerInputs>,
    effects: Res<AudioChannel<EffectsChannel>>,
    element_assets: Res<Assets<MapElementMeta>>,
    collision_world: CollisionWorld,
) {
    // Helper to spawn damage regions
    let mut spawn_damage_region =
        |commands: &mut Commands, pos: Vec2, size: Vec2, owner: Entity| {
            commands
                .spawn()
                .insert(Rollback::new(ridp.next_id()))
                .insert(Transform::from_translation(pos.extend(0.0)))
                .insert(GlobalTransform::default())
                .insert(DamageRegion { size })
                .insert(DamageRegionOwner(owner))
                .insert(Lifetime::new(2.0 / 60.0));
        };

    for (
        item_ent,
        mut transform,
        mut sword,
        mut sprite,
        mut body,
        meta_handle,
        parent,
        item_used,
        item_dropped,
    ) in &mut swords
    {
        let meta = element_assets.get(meta_handle).unwrap();
        let BuiltinElementKind::Sword { sound_handle, angular_velocity, throw_velocity, arm_delay, .. } = &meta.builtin else {
            unreachable!();
        };

        // For all tiems that are being held
        if let Some(parent) = parent {
            let (player_sprite, player_transform, ..) =
                players.get(parent.get()).expect("Parent is not a player");

            // Deactivate collisions while being held
            body.is_deactivated = true;

            // Flip the sprite to match the player orientation
            let flip = player_sprite.flip_x;
            sprite.flip_x = flip;
            let flip_factor = if flip { -1.0 } else { 1.0 };
            transform.translation.x = 13.0 * flip_factor;
            transform.translation.y = 21.0;
            transform.translation.z = 0.0;
            transform.rotation = Quat::IDENTITY;

            // Reset the sword animation if we're not swinging it
            if !matches!(sword.state, SwordState::Swinging { .. }) {
                sprite.start = 4;
                sprite.end = 4;
                sprite.index = 0;
                sprite.repeat = false;
            }

            let mut next_state = None;
            match &mut sword.state {
                SwordState::Idle => (),
                SwordState::Swinging { frame } => {
                    // If we're at the end of the swinging animation
                    if sprite.index >= sprite.end.saturating_sub(sprite.start).saturating_sub(1) {
                        // Go to cooldown frames
                        next_state = Some(SwordState::Cooldown { frame: 0 });

                    // If we're still swinging
                    } else {
                        // Set the current attack frame to the animation index
                        *frame = sprite.index;
                    }

                    // TODO: Move all these constants to the builtin item config
                    match frame {
                        0 => spawn_damage_region(
                            &mut commands,
                            Vec2::new(
                                player_transform.translation.x + 20.0 * flip_factor,
                                player_transform.translation.y + 20.0,
                            ),
                            Vec2::new(30.0, 70.0),
                            parent.get(),
                        ),
                        1 => spawn_damage_region(
                            &mut commands,
                            Vec2::new(
                                player_transform.translation.x + 25.0 * flip_factor,
                                player_transform.translation.y + 20.0,
                            ),
                            Vec2::new(40.0, 50.0),
                            parent.get(),
                        ),
                        2 => spawn_damage_region(
                            &mut commands,
                            Vec2::new(
                                player_transform.translation.x + 20.0 * flip_factor,
                                player_transform.translation.y,
                            ),
                            Vec2::new(40.0, 50.0),
                            parent.get(),
                        ),
                        _ => (),
                    }

                    *frame += 1;
                }
                SwordState::Cooldown { frame } => {
                    if *frame >= COOLDOWN_FRAMES {
                        next_state = Some(SwordState::Idle);
                    } else {
                        *frame += 1;
                    }
                }
            }

            if let Some(next) = next_state {
                sword.state = next;
            }
            if item_used.is_some() {
                commands.entity(item_ent).remove::<ItemUsed>();
            }

            // If the item is being used
            if item_used.is_some() && matches!(sword.state, SwordState::Idle) {
                sprite.index = 0;
                sprite.start = 8;
                sprite.end = 12;
                sword.state = SwordState::Swinging { frame: 0 };
                if player_inputs.is_confirmed {
                    effects.play(sound_handle.clone_weak());
                }
            }
        } else {
            sword.dropped_time += 1.0 / crate::FPS as f32;
        }

        // If the item was dropped
        if let Some(dropped) = item_dropped {
            sword.dropped_time = 0.0;

            commands.entity(item_ent).remove::<ItemDropped>();
            let (.., player_transform, player_body) =
                players.get(dropped.player).expect("Parent is not a player");

            // Re-activate physics
            body.is_deactivated = false;

            // Put sword in rest position
            sprite.index = 0;
            sprite.start = 0;
            sprite.end = 0;

            if player_body.velocity != Vec2::ZERO {
                let horizontal_flip_factor = if sprite.flip_x {
                    Vec2::new(-1.0, 1.0)
                } else {
                    Vec2::ONE
                };
                body.velocity = *throw_velocity * horizontal_flip_factor + player_body.velocity;
                body.angular_velocity = *angular_velocity * if sprite.flip_x { -1.0 } else { 1.0 };
            }
            body.is_spawning = true;

            transform.translation.y = player_transform.translation.y;
            transform.translation.x = player_transform.translation.x;
            transform.translation.z = player_transform.translation.z;
        }

        if parent.is_none() && body.velocity != Vec2::ZERO && sword.dropped_time >= *arm_delay {
            for player in collision_world
                .actor_collisions(item_ent)
                .into_iter()
                .filter(|&x| players.contains(x))
            {
                commands.add(PlayerKillCommand::new(player));
            }
        }
    }
}
